src/v/kafka/server/handlers/fetch.cc

```diff
@@ -805,4 +929,11 @@ void op_context::response_placeholder::set(
     }
 }
 
+size_t fetch_memory_estimator(
+  const size_t request_size, connection_context& /*conn_ctx*/) {
+    return request_size
+             * 3 // appx memory for fetch plans, fetch configs, read results
+           + batch_size_estimate; // at least one batch of the data will be read
```

Comment:
In principle I like the idea of reserving this  1MiB up-front, but trying to think about the limiting cases.  If someone has like 2GiB of memory per shard, then the Kafka 30% of that is 600MiB, so our practical limit on how many concurrent fetches could happen per shard is 600.

If  we had e.g. a 3 node 4 core system, that would be 7200 concurrent fetches (assuming perfect distribution), so it's probably enough for a system with our ~5000 client limit, but it's close.

This probably doesn't matter if the fetches are completing and returning promptly, but I think in some cases we do a kind of long-polling on fetches?  That would be the situation where the static 1MiB minimum units could cause an issue.

This may well still be the right thing to do, but we should make sure we're aligning our stated limits on client counts etc with what works based on this memory limit.

Reply:
Besides the great point on the stated limits, I'm still undecided whether `batch_size_estimate` should be configurable. That way a customer will be able to build a more consume-oriented cluster by knowingly limiting batch sizes to a smaller value, and by configuring this estimated size accordingly. WDYT?

Reply:
>but I think in some cases we do a kind of long-polling on fetches? 

good point. perhaps we could give back the units during the long poll sleep, and "restart" that resource acquisition on wake up? also I seem to recall that the long poll isn't indefinite--there is a bound at which point we return to the client empty handed.